libs/utils/src/sync/gate.rs

```diff
@@ -88,11 +106,20 @@ impl Gate {
     /// to avoid blocking close() indefinitely: typically types that contain a Gate will
     /// also contain a CancellationToken.
     pub fn enter(&self) -> Result<GateGuard, GateError> {
-        self.sem
-            .clone()
-            .try_acquire_owned()
-            .map(GateGuard)
-            .map_err(|_| GateError::GateClosed)
+        let permit = self
+            .inner
+            .sem
+            .try_acquire()
+            .map_err(|_| GateError::GateClosed)?;
+
+        // we now have the permit, let's disable the normal raii functionality and leave
```

Comment:
Would be good to say _why_ we're avoiding RAII (to avoid the extra Arc<> since we have our own Arc to the gate)

Reply:
I'll add the why. a90f84246